go/nbs/dynamo_table_reader.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2017 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package nbs
+
+import (
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/attic-labs/noms/go/d"
+	"github.com/attic-labs/noms/go/util/sizecache"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+)
+
+const (
+	dataAttr    = "data"
+	tablePrefix = "*" // I want to use NBS table names as keys when they are written to DynamoDB, but a bare table name is a legal Noms Database name as well. To avoid collisions, dynamoTableReader prepends this prefix (which is not a legal character in a Noms Database name).
+)
+
+// dynamoTableReader assumes the existence of a DynamoDB table whose primary partition key is in String format and named `db`.
+type dynamoTableReader struct {
+	tableReader
+	ddb   ddbsvc
+	table string
+	h     addr
+	tc    *sizecache.SizeCache
+}
+
+type tableNotInDynamoErr struct {
+	nbs, dynamo string
+}
+
+func (t tableNotInDynamoErr) Error() string {
+	return fmt.Sprintf("NBS table %s not present in DynamoDB table %s", t.nbs, t.dynamo)
+}
+
+func newDynamoTableReader(ddb ddbsvc, table string, h addr, chunkCount uint32, data []byte, indexCache *indexCache, tc *sizecache.SizeCache) chunkSource {
+	d.PanicIfTrue(table == "")
+	source := &dynamoTableReader{ddb: ddb, table: table, h: h, tc: tc}
+
+	var index tableIndex
+	found := false
+	if indexCache != nil {
+		indexCache.lockEntry(h)
+		defer indexCache.unlockEntry(h)
+		index, found = indexCache.get(h)
+	}
+
+	if !found {
+		index = parseTableIndex(data)
+		if indexCache != nil {
+			indexCache.put(h, index)
+		}
+	}
+
+	source.tableReader = newTableReader(index, source, fileBlockSize)
+	d.PanicIfFalse(chunkCount == source.count())
+	return source
+}
+
+func (dtr *dynamoTableReader) hash() addr {
+	return dtr.h
+}
+
+func (dtr *dynamoTableReader) ReadAtWithStats(p []byte, off int64, stats *Stats) (n int, err error) {
+	if data, present := dynamoTableCacheMaybeGet(dtr.tc, dtr.hash()); present {
+		n = copy(p, data[off:])
```

Comment:
since this is essentially the start of the L1 (mem) portion of the table cache, can we add `MemBytesPerRead` and `MemReadLatency` stats to start collecting. That way we can still compare apples-to-apples when reasoning about total read load, regardless of where the reads ended up